Sort Sites Dashboard by Last Publish

wojtekn opened the pull request:
#### Proposed Changes

We've already added the Last Publish column that uses the site's `updated_at` field to display nicely formatted dates.

In this PR I've added a hook that sorts sites using the Last Publish field.

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1. Load Sites Dashboard at `/sites-dashboard` URL
2. Confirm that sites are sorted by Last Publish date, descending - the latest updated site should be the first one.
3. Confirm that sorting works when we switch tabs and when we filter the list using search phrase.

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in [Simple](P9HQHe-k8-p2-p2), [Atomic](P9HQHe-jW-p2-p2), and [self-hosted Jetpack sites](PCYsg-g6b-p2-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Are we memoizing when appropriate (for expensive computations)? More info in [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md) and [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors)
- [ ] ~~Have we sent any new strings [for translation](PCYsg-1vr-p2-p2) ASAP?~~

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/wp-calypso/issues/65168

wojtekn committed:
Sort Sites Dashboard by Last Publish

wojtekn committed:
Fix asc/desc logic as now it works inversely

wojtekn committed:
Disable fuzzy search sorting as we want to control the sort order

wojtekn committed:
Remove fuzzy search change to ensure it doesn't affect other places

danielbachhuber commented on client/sites-dashboard/components/sites-dashboard.tsx:
```diff
@@ -89,6 +95,11 @@ export function SitesDashboard( { queryParams: { search, status } }: SitesDashbo
 	const selectedTabName = tabs.find( ( tab ) => tab.name === status )?.name;
 
+	const { sortedSites } = useSitesTableSorting( filteredSites, {
+		sort: 'last-publish',
```
What do you think about calling this `updated-at`?

My personal preference is to name query params, etc. after the data they're referencing, vs. UI. The UI may change, but the underlying data won't. It also makes it more obvious which data the query param is referring to.

wojtekn replied:
Thanks, it's a good point, it makes perfect sense. I've changed that to `updated-at`.

danielbachhuber commented on packages/components/src/sites-table/use-sites-table-sorting.tsx:
```diff
@@ -0,0 +1,35 @@
+// eslint-disable-next-line no-restricted-imports
```
Does it make sense to have some tests for the various types of data inputs this selector will support?

wojtekn replied:
Do you mean to validate `sort` parameter values and handle incorrect ones somehow? Or to add tests for the sorting feature?

danielbachhuber replied:
More the latter: tests for the sorting feature, and different values of data it can sort.

wojtekn replied:
Thanks, I've added tests for a few cases.

danielbachhuber commented on client/sites-dashboard/components/sites-dashboard.tsx:
```diff
@@ -118,8 +129,8 @@ export function SitesDashboard( { queryParams: { search, status } }: SitesDashbo
 								</SearchWrapper>
-								{ filteredSites.length > 0 ? (
-									<SitesTable sites={ filteredSites } />
+								{ sortedSites.length > 0 ? (
```
From a product perspective, sorting by `updatedAt` is much less useful than having my primary and most recently visited sites at the top of the list.

Maybe we should make sorting by `updatedAt` opt-in behavior?

wojtekn replied:
We wanted to sort by Last Publish in the initial iteration (https://github.com/Automattic/wp-calypso/issues/65168).

Letting a user decide which field we should use to sort (e.g., by clicking on column names), what should be the sort order (ascending or descending), and what should be the default sort order makes sense, but I think it deserves the separate research and discussion. What do you think?

danielbachhuber replied:
> What do you think?

I still think: _From a product perspective, sorting by `updatedAt` is much less useful than having my primary and most recently visited sites at the top of the list._

I'd be fine to land this PR as-is, though.

p-jackson replied:
[I feel the same](https://github.com/Automattic/wp-calypso/issues/65168#issuecomment-1196518433), but agree that we can merge while still continuing that conversation.

danielbachhuber reviewed (commented):
👍 Code looks good, left a couple of comments.

Out of habit, I merged `trunk` into this branch, so you'll want to pull before any changes.

wojtekn committed:
Rename filtering constant to make it closer to the data it uses

wojtekn commented:
> Out of habit, I merged trunk into this branch, so you'll want to pull before any changes.

Thanks for reviewing it @danielbachhuber. I've just rebased my local branch upon the trunk and pushed it as this is a recommended workflow for Calypso.

vykes-mac commented:
Tested it an it works, however the 'Last Publish' column and the information displayed feels like two different things. We might need to refine this further to accurately convey to the user what is intended.

wojtekn committed:
Add tests for sites sorting hook

p-jackson commented on packages/components/src/sites-table/use-sites-table-sorting.tsx:
```diff
@@ -0,0 +1,35 @@
+import type { SiteExcerptData } from 'calypso/data/sites/site-excerpt-types';
+
+interface SitesTableSortOptions {
+	sort?: string;
```
Maybe `sortKey`?

```suggestion
	sortKey?: 'updated-at';
```

wojtekn replied:
That makes sense. I renamed `order` to `sortOrder` to keep the naming convention consistent.

p-jackson commented on packages/components/src/sites-table/use-sites-table-sorting.tsx:
```diff
@@ -0,0 +1,35 @@
+
+interface SitesTableSortOptions {
+	sort?: string;
+	order?: string;
```
If we only support two possible orderings, then this is a good way to document and enforce that. It will probably make one of the unit tests (and maybe the `switch` statement?) redundant because Typescript will know those things aren't possible.

```suggestion
	order?: 'asc' | 'desc';
```

wojtekn replied:
I adjusted it to define two allowed options. When I try to set order to not supported option, I get a TS warning in IDE. However, it doesn't break compilation, and it doesn't throw any error in the browser console. Do you think it would be wise to leave the test in place in this case?

p-jackson replied:
Personally I wouldn't test using a value that TS is already warning about, but I don't think there's anything necessarily wrong with it. _Technically_ it's still possible that an unsupported could slip in at runtime.
But I think the whole point of TS is to prohibit a whole range of possibilities so that you don't have to think about or test for them. For instance we don't write unit tests so that we get well defined behaviour when `sortOrder` is a number or an `instanceof MyClass`. And I think of an invalid string (that TS can check for) as just another such invalid argument.

wojtekn replied:
It makes sense, I've removed it then.

p-jackson commented on packages/components/src/sites-table/use-sites-table-sorting.tsx:
```diff
@@ -0,0 +1,35 @@
+	return sites.sort( ( a, b ) => {
+		if ( a.options?.updated_at && b.options?.updated_at ) {
+			return (
+				( Date.parse( a.options?.updated_at ) - Date.parse( b.options?.updated_at ) ) *
```
The `updated_at` field uses the `YYYY-MM-DDTHH:MM:SS` format, which means it's sortable using regular `<` or `>` comparisons. I like the trick of multiplying by `1` or `-1`, but I suspect that calling the `Date.parse()` function is much more expensive than doing string comparisons.

Also, not a biggy, but the `?` aren't necessary here because we've already checked whether the property exists in the `if` statement above, and TypeScript is clever enough to figure that out.

```suggestion
				( Date.parse( a.options.updated_at ) - Date.parse( b.options.updated_at ) ) *
```

wojtekn replied:
@p-jackson, thanks for the suggestions. I've updated the sorting implementation to use string comparison.

p-jackson replied:
Probably only thought of it because of today's discussion on dates 😸  p1658927993059409-slack-C025Q5RK2
I'm not sure we have to account for leap-second smearing though 🤔

p-jackson commented on client/sites-dashboard/components/sites-dashboard.tsx:
```diff
@@ -89,6 +95,11 @@ export function SitesDashboard( { queryParams: { search, status } }: SitesDashbo
 
 	const selectedTabName = tabs.find( ( tab ) => tab.name === status )?.name;
 
+	const { sortedSites } = useSitesTableSorting( filteredSites, {
```
This is running the sorting after the filtering, but I think it could actually be more efficient to do the sorting first. That way we will only resort when the data from the query changes rather than every time the user changes the filtering.

wojtekn replied:
@p-jackson as far as I see, the whole logic (useSiteExcerptsQuery, useSitesTableFiltering) is executed on each tab change and on each search query change. If we do sorting first, wouldn't we sort the whole dataset each time?

p-jackson replied:
If sorting was done first then on each tab change:
- `useSiteExcerptsQuery` is called, but the underlying `useQuery` hook will fetch data from cache so this is quick
- then `useSitesTableSorting` is called, but since the result from `useSiteExcerptsQuery` hasn't changed `useMemo`  will return the pre-sorted data from memory
- then `useSitesTableFiltering` is called, but since `useSitesTableSorting` is returning an identical set of sites it doesn't retrigger filtering and the `useMemo` that does the filtering will return the exact same pre-filtered set of sites.

(this is assuming `useSitesTableSorting` is using `useMemo` under the hood)

wojtekn replied:
I see, I adjusted that. I needed to disable fuzzy search sorting too as it was sorting records by relevance.

p-jackson replied:
Hmmm I think we might need that fuzzy search sorting to come back. It's still good to run the `useSitesTableSorting` hook first because it saves on sorting when something else (other than searching) has to re-render. But I find the fuzzy search is quite counterintuitive without sorting by relevance. Sometimes it means that a site that has very little to do with my search term appears at the top of the list, even when there's a perfect match available in the table.

p-jackson commented on packages/components/src/sites-table/use-sites-table-sorting.tsx:
```diff
@@ -0,0 +1,35 @@
+	allSites: SiteExcerptData[],
+	{ sort, order = 'asc' }: SitesTableSortOptions
+): UseSitesTableSortingResult {
+	switch ( sort ) {
+		case 'updated-at':
+			return { sortedSites: sortSitesByLastPublish( allSites, order ) };
+		default:
+			return { sortedSites: allSites };
+	}
```
Maybe we should wrap this in a `useMemo` so that we only resort when it's actually necessary.

wojtekn replied:
Thanks, I wrapped it with the `useMemo`.

p-jackson reviewed (commented):
Good to get some explicit sorting implemented. As you know, I'm not convinced that sorting exclusively by `updated_at` is the right thing to do. But I doubt we'll leave the sorting as-is either, so it's good to implement the hook we will need.

I'm also wondering whether `useSitesTableSorting` should be a separate hook, or whether it should just be another feature of the `useSitesTableFiltering` hook. I mention in a comment below that "sorting then filtering" is going to be more efficient than "filtering then sorting", and if it were all combined in one hook it would always be done in the most efficient way. On the other hand separation of concerns is good.

I haven't managed to convince myself that one way is better than the other 🤷

wojtekn committed:
Speed up sorting by using string comparison instead of parsed dates

wojtekn committed:
Make order option values more strict

wojtekn committed:
Rename sort and order options to make them more meaningful

wojtekn commented:
> I'm also wondering whether useSitesTableSorting should be a separate hook, or whether it should just be another feature of the useSitesTableFiltering hook. I mention in a comment below that "sorting then filtering" is going to be more efficient than "filtering then sorting", and if it were all combined in one hook it would always be done in the most efficient way. On the other hand separation of concerns is good.

> I haven't managed to convince myself that one way is better than the other 🤷

@p-jackson, thanks for the feedback. @zaguiini also brought that point. If we want to move a sorting inside the filter hook, we could even try to use `sortFn` option of the Fuzzy Search component and provide a sorting hook there. 

I couldn't see any advantage of doing that, though. With the current approach, we have a search executed in the filtering component and the sorting done later. I see the pros:

* sorting doesn’t depend on searching. We can replace search solution implementation, and sorting won’t be affected
* the chain of command is shorter and easier to understand, it's more aligned with separation of concepts and single responsibility design principles
* we provide an entry point to implement URL-based sorting (e.g., to allow users to sort table content by clicking on the headers) directly in the `sites-dashboard.tsx` - as it accepts sort and order params there.
* the sorting component is prepared to add support for other types of sort (e.g., name and remaining fields displayed in the table).

wojtekn committed:
Memoize sorted sites data

p-jackson commented:
I think your rationale for keeping the hooks separate for now makes sense.

The reason the tests fail (but work fine when manually testing) is that `useMemo` has an assertion to ensure it's always being used within a React context.

This is a good example of how hooks can be tested in a React environment.
https://github.com/Automattic/wp-calypso/blob/12c70af98059fe1dd9dee93d4b09d01170d4d3b2/client/landing/stepper/hooks/test/use-site-id-param.ts#L18

It uses `@testing-library/react-hooks`. It uses a wrapper which means it's not as simple as calling the hook like a regular function like we were doing before. But on the other hand the testing library allows you to try all sorts of interesting things that hooks can do like having different behaviour the first time it's run vs. if you call it again just with slightly different arguments. Not necessarily needed in this case (unless you wanted to test that the hook was definitely memoising correctly), but it's usually how hooks need to be tested.

wojtekn committed:
Remove test that is no longer needed atfer making types stricter

wojtekn committed:
Fix tests to ensure they work after adding useMemo

useMemo asserts to ensure it is always used withing React context,
and tests started failing after introducing memoizing. I'm rewriting
tests to use the renderHook from React Testing Library to accoount
for that.

wojtekn committed:
Move sorting before filtering

I'm moving sorting to earlier phase to sort full data set once,
memoize it, and then filters should operate on cached data.

p-jackson commented on packages/components/package.json:
```diff
@@ -53,6 +53,7 @@
 		"@storybook/react": "^6.4.18",
 		"@testing-library/jest-dom": "^5.16.2",
 		"@testing-library/react": "^12.1.3",
+		"@testing-library/react-hooks": "^8.0.1",
```
The rest of the monorepo is using `7.0.2` (tested with `npm ls @testing-library/react-hooks`) so I'll update the PR to use the older version. [There isn't a significant difference between the versions](https://github.com/testing-library/react-hooks-testing-library/releases).

p-jackson commented on packages/components/src/sites-table/use-sites-table-filtering.tsx:
```diff
@@ -27,6 +27,9 @@ export function useSitesTableFiltering(
 		keys: [ 'URL', 'name', 'slug' ],
 		query: search,
+		options: {
+			shouldSort: false,
```
[As I mentioned here](https://github.com/Automattic/wp-calypso/pull/65882#discussion_r932793457), I'm going to remove this option.

p-jackson committed:
Change react-hooks version to match the rest of monorepo

p-jackson committed:
Typo

p-jackson committed:
Re-instate fuzzy search ordering

Fuzzy search leaves items in the table even when they don't technically
match the search term. This means sorting by relevance is an important
part of the fuzzy search experience. Otherwise you can end up with a
perfect match appearing at the bottom of the list, and a row that has
very little to do with the search term appearing at the top.

p-jackson reviewed (approved):
Thanks for the updates. I've made some updates and am going to merge soon.

sirbrillig commented on packages/components/src/sites-table/use-sites-table-sorting.tsx:
```diff
@@ -0,0 +1,44 @@
+import { useMemo } from 'react';
+// eslint-disable-next-line no-restricted-imports
+import type { SiteExcerptData } from 'calypso/data/sites/site-excerpt-types';
```
A late drive-by comment but is this a good idea? This imports data from the `calypso` package into the `@automattic/components` package, which is in turn imported into the `calypso` package, creating a circular dependency. It also adds a dependency to the `@automattic/components` package which is not part of the `dependencies` array in package.json.

I think this should be removed. If you need `SiteExcerptData` in this package, you should probably duplicate it here or move it to another package which `@automattic/components` can depend on.

sirbrillig replied:
I've attempted a fix for this here: https://github.com/Automattic/wp-calypso/pull/66778

p-jackson replied:
Yes, very valid point.

We've already changed our other heavy-lifting hook (`useSitesTableFiltering()`) so that it operates on generic data (anything that looks sorta like a site), but forgot to do the same thing for this hook.
https://github.com/Automattic/wp-calypso/blob/2d6bf7416c800955578f831a3131fb1c407db6f1/packages/components/src/sites-table/use-sites-table-filtering.tsx#L45

I'll take a look at #66778
